Could not find file 'C:\Program Files\dotnet\sdk\1.0.4\minimumMSBuildVersion'

@vsccarl commented on [Fri May 26 2017](https://github.com/dotnet/corefx/issues/20386)

In the latest d15prerel (26526.3001) I'm getting the below error when trying to build any .net Core project. The logs are the same for both C# and VB. The folder exists, but as the logs say, the file is missing.

```
1>------ Build started: Project: ConsoleApp3, Configuration: Debug Any CPU ------
1>C:\Users\v-sccarl\Source\Repos\ConsoleApp3\ConsoleApp3\ConsoleApp3.vbproj : error MSB4014: The build stopped unexpectedly because of an internal failure.
1>C:\Users\v-sccarl\Source\Repos\ConsoleApp3\ConsoleApp3\ConsoleApp3.vbproj : error MSB4014: System.IO.FileNotFoundException: Could not find file 'C:\Program Files\dotnet\sdk\1.0.4\minimumMSBuildVersion'.
1>C:\Users\v-sccarl\Source\Repos\ConsoleApp3\ConsoleApp3\ConsoleApp3.vbproj : error MSB4014: File name: 'C:\Program Files\dotnet\sdk\1.0.4\minimumMSBuildVersion'
1>C:\Users\v-sccarl\Source\Repos\ConsoleApp3\ConsoleApp3\ConsoleApp3.vbproj : error MSB4014:    at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
1>C:\Users\v-sccarl\Source\Repos\ConsoleApp3\ConsoleApp3\ConsoleApp3.vbproj : error MSB4014:    at System.IO.FileStream.Init(String path, FileMode mode, FileAccess access, Int32 rights, Boolean useRights, FileShare share, Int32 bufferSize, FileOptions options, SECURITY_ATTRIBUTES secAttrs, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
1>C:\Users\v-sccarl\Source\Repos\ConsoleApp3\ConsoleApp3\ConsoleApp3.vbproj : error MSB4014:    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
1>C:\Users\v-sccarl\Source\Repos\ConsoleApp3\ConsoleApp3\ConsoleApp3.vbproj : error MSB4014:    at System.IO.StreamReader..ctor(String path, Encoding encoding, Boolean detectEncodingFromByteOrderMarks, Int32 bufferSize, Boolean checkHost)
1>C:\Users\v-sccarl\Source\Repos\ConsoleApp3\ConsoleApp3\ConsoleApp3.vbproj : error MSB4014:    at System.IO.StreamReader..ctor(String path, Encoding encoding)
1>C:\Users\v-sccarl\Source\Repos\ConsoleApp3\ConsoleApp3\ConsoleApp3.vbproj : error MSB4014:    at System.IO.File.InternalReadAllLines(String path, Encoding encoding)
1>C:\Users\v-sccarl\Source\Repos\ConsoleApp3\ConsoleApp3\ConsoleApp3.vbproj : error MSB4014:    at System.IO.File.ReadAllLines(String path)
1>C:\Users\v-sccarl\Source\Repos\ConsoleApp3\ConsoleApp3\ConsoleApp3.vbproj : error MSB4014:    at Microsoft.DotNet.MSBuildSdkResolver.DotNetMSBuildSdkResolver.Resolve(SdkReference sdkReference, SdkResolverContext context, SdkResultFactory factory)
1>C:\Users\v-sccarl\Source\Repos\ConsoleApp3\ConsoleApp3\ConsoleApp3.vbproj : error MSB4014:    at Microsoft.Build.BackEnd.SdkResolution.GetSdkPath(SdkReference sdk, ILoggingService logger, BuildEventContext buildEventContext, ElementLocation sdkReferenceLocation, String solutionPath, String projectPath)
1>C:\Users\v-sccarl\Source\Repos\ConsoleApp3\ConsoleApp3\ConsoleApp3.vbproj : error MSB4014: The build stopped unexpectedly because of an internal failure.
1>C:\Users\v-sccarl\Source\Repos\ConsoleApp3\ConsoleApp3\ConsoleApp3.vbproj : error MSB4014: System.IO.FileNotFoundException: Could not find file 'C:\Program Files\dotnet\sdk\1.0.4\minimumMSBuildVersion'.
1>C:\Users\v-sccarl\Source\Repos\ConsoleApp3\ConsoleApp3\ConsoleApp3.vbproj : error MSB4014: File name: 'C:\Program Files\dotnet\sdk\1.0.4\minimumMSBuildVersion'
1>C:\Users\v-sccarl\Source\Repos\ConsoleApp3\ConsoleApp3\ConsoleApp3.vbproj : error MSB4014:    at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
1>C:\Users\v-sccarl\Source\Repos\ConsoleApp3\ConsoleApp3\ConsoleApp3.vbproj : error MSB4014:    at System.IO.FileStream.Init(String path, FileMode mode, FileAccess access, Int32 rights, Boolean useRights, FileShare share, Int32 bufferSize, FileOptions options, SECURITY_ATTRIBUTES secAttrs, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
1>C:\Users\v-sccarl\Source\Repos\ConsoleApp3\ConsoleApp3\ConsoleApp3.vbproj : error MSB4014:    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
1>C:\Users\v-sccarl\Source\Repos\ConsoleApp3\ConsoleApp3\ConsoleApp3.vbproj : error MSB4014:    at System.IO.StreamReader..ctor(String path, Encoding encoding, Boolean detectEncodingFromByteOrderMarks, Int32 bufferSize, Boolean checkHost)
1>C:\Users\v-sccarl\Source\Repos\ConsoleApp3\ConsoleApp3\ConsoleApp3.vbproj : error MSB4014:    at System.IO.StreamReader..ctor(String path, Encoding encoding)
1>C:\Users\v-sccarl\Source\Repos\ConsoleApp3\ConsoleApp3\ConsoleApp3.vbproj : error MSB4014:    at System.IO.File.InternalReadAllLines(String path, Encoding encoding)
1>C:\Users\v-sccarl\Source\Repos\ConsoleApp3\ConsoleApp3\ConsoleApp3.vbproj : error MSB4014:    at System.IO.File.ReadAllLines(String path)
1>C:\Users\v-sccarl\Source\Repos\ConsoleApp3\ConsoleApp3\ConsoleApp3.vbproj : error MSB4014:    at Microsoft.DotNet.MSBuildSdkResolver.DotNetMSBuildSdkResolver.Resolve(SdkReference sdkReference, SdkResolverContext context, SdkResultFactory factory)
1>C:\Users\v-sccarl\Source\Repos\ConsoleApp3\ConsoleApp3\ConsoleApp3.vbproj : error MSB4014:    at Microsoft.Build.BackEnd.SdkResolution.GetSdkPath(SdkReference sdk, ILoggingService logger, BuildEventContext buildEventContext, ElementLocation sdkReferenceLocation, String solutionPath, String projectPath)
1>vbc : warning BC40057: Namespace or type specified in the project-level Imports 'System.Data' doesn't contain any public member or cannot be found. Make sure the namespace or the type is defined and contains at least one public member. Make sure the imported element name doesn't use any aliases.
1>ConsoleApp3 -> C:\Users\v-sccarl\Source\Repos\ConsoleApp3\ConsoleApp3\bin\Debug\netcoreapp1.1\ConsoleApp3.dll
1>Done building project "ConsoleApp3.vbproj".
========== Build: 1 succeeded or up-to-date, 0 failed, 0 skipped ==========

```

--- a/src/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs
+++ b/src/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs
@@ -5,6 +5,7 @@ using Microsoft.Build.Framework;
 using System;
 using System.Collections.Generic;
 using System.IO;
+using System.Linq;
 
 namespace Microsoft.DotNet.MSBuildSdkResolver
 {
@@ -63,9 +64,7 @@ namespace Microsoft.DotNet.MSBuildSdkResolver
                             + " the version specified in global.json."
                         });
                 }
-
-                var minimumMSBuildVersionString =
-                    File.ReadAllLines(Path.Combine(netcoreSdkDir, "minimumMSBuildVersion"))[0];
+                string minimumMSBuildVersionString = GetMinimumMSBuildVersion(netcoreSdkDir);
                 var minimumMSBuildVersion = Version.Parse(minimumMSBuildVersionString);
                 if (context.MSBuildVersion < minimumMSBuildVersion)
                 {
@@ -92,6 +91,20 @@ namespace Microsoft.DotNet.MSBuildSdkResolver
             }
 
             return factory.IndicateSuccess(msbuildSdkDir, netcoreSdkVersion);
+        }
+
+        private static string GetMinimumMSBuildVersion(string netcoreSdkDir)
+        {
+            string minimumVersionFilePath = Path.Combine(netcoreSdkDir, "minimumMSBuildVersion");
+            if (!File.Exists(minimumVersionFilePath))
+            {
+                // smallest version that had resolver support and also
+                // greater than or equal to the version required by any 
+                // .NET Core SDK that did not have this file.
+                return "15.3.0";  
+            }
+
+            return File.ReadLines(minimumVersionFilePath).First().Trim();
         }
 
         private bool IsNetCoreSDKSmallerThanTheMinimumVersion(string netcoreSdkVersion, string minimumVersion)

--- a/test/Microsoft.DotNet.MSBuildSdkResolver.Tests/GivenAnMSBuildSdkResolver.cs
+++ b/test/Microsoft.DotNet.MSBuildSdkResolver.Tests/GivenAnMSBuildSdkResolver.cs
@@ -91,6 +91,7 @@ namespace Microsoft.DotNet.Cli.Utils.Tests
                 new SdkReference("Some.Test.Sdk", null, "99.99.99"),
                 new MockContext
                 {
+                    MSBuildVersion = new Version(1, 0),
                     ProjectFilePath = environment.TestDirectory.FullName
                 },
                 new MockFactory());
@@ -194,7 +195,10 @@ namespace Microsoft.DotNet.Cli.Utils.Tests
                 var dir = GetSdkDirectory(programFiles, sdkName, sdkVersion);
                 dir.Create();
 
-                CreateMSBuildRequiredVersionFile(programFiles, sdkVersion, minimumMSBuildVersion);
+                if (minimumMSBuildVersion != null)
+                {
+                    CreateMSBuildRequiredVersionFile(programFiles, sdkVersion, minimumMSBuildVersion);
+                }
 
                 return dir;
             }
@@ -254,7 +258,7 @@ namespace Microsoft.DotNet.Cli.Utils.Tests
 
             public MockContext()
             {
-                MSBuildVersion = new Version(1, 0);
+                MSBuildVersion = new Version(15, 3, 0);
             }
         }
 

